sysboundary/sysboundarycondition.cpp

```diff
@@ -785,48 +785,55 @@ namespace SBC {
       return flowtoCellsBlock;
    }
    
-   Real SysBoundaryCondition::fieldBoundaryCopyFromExistingFaceNbrMagneticField(
+   Real SysBoundaryCondition::fieldBoundaryCopyFromSolvingNbrMagneticField(
       FsGrid< std::array<Real, fsgrids::bfield::N_BFIELD>, 2> & perBGrid,
       FsGrid< fsgrids::technical, 2> & technicalGrid,
       cint i,
       cint j,
       cint k,
-      cuint component
+      cuint component,
+      cuint mask
    ) {
-      const std::array<int,3> closestCell = getTheClosestNonsysboundaryCell(technicalGrid, i, j, k);
-      
-      const std::array<int32_t, 3> gid = technicalGrid.getGlobalIndices(i, j, k);
-      const std::array<int32_t, 3> ngid = technicalGrid.getGlobalIndices(closestCell[0], closestCell[1], closestCell[2]);
 
-      if (closestCell[0] == std::numeric_limits<int>::min()) {
-         //cerr << "(" << gid[0] << "," << gid[1] << "," << gid[2] << ")" << __FILE__ << ":" << __LINE__ << ": No closest cell found!" << endl;
-         //abort();
+      int distance = std::numeric_limits<int>::max();
+      std::vector< std::array<int,3> > closestCells;
 
-         // When mpiGrid is refined, the fsgrid boundary layer has a width greater than 2. In this case,
-         // the boundary cells that do not find a non-boundary neighbor just keep their original value,
-         // we don't care what happens in them since they have no effect on the Vlasov solver.
-         return perBGrid.get(i,j,k)->at(fsgrids::bfield::PERBX+component);
+      for (int kk=-2; kk<3; kk++) {
+         for (int jj=-2; jj<3; jj++) {
+            for (int ii=-2; ii<3 ; ii++) {
+               if( technicalGrid.get(i+ii,j+jj,k+kk) // skip invalid cells returning NULL
+                   && (technicalGrid.get(i+ii,j+jj,k+kk)->SOLVE & mask) == mask // Did that guy solve this component?
+                   && technicalGrid.get(i+ii,j+jj,k+kk)->sysBoundaryFlag != sysboundarytype::DO_NOT_COMPUTE // Do not copy from there
+               ) {
+                  distance = min(distance, ii*ii + jj*jj + kk*kk);
```

Comment:
This would consolidate these into one loop:

before loops:
std::array<int, 3> ClosestCell = {-1,-1,-1};
delete closestcells vector

here:
int d = ii*ii + jj*jj + kk*kk;
if (distance > d) {
   ClosestCell = {i+ii, j+jj, k+kk};
   distance = d;
}

Below:
if ((closestCell[0]<0) || (closestCell[1]<0) || (closestCell[0]<0)) {
         cerr << __FILE__ << ":" << __LINE__ << ": No closest cell found!" << endl;
         abort();
      }

      return perBGrid.get(closestCell[0], closestCell[1], closestCell[2])->at(fsgrids::bfield::PERBX+component);